src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1427,6 +1561,69 @@ private bool HasParamsCollectionTypeInProgress(NamedTypeSymbol toCheck,
             return false;
         }
 
+        internal bool HasParamsCollectionTypeApplicableIndexerInProgress(SyntaxNode syntax, TypeSymbol targetType)
+        {
+            Binder? current = this;
+            do
+            {
+                if (current is ParamsCollectionTypeApplicableIndexerInProgress binder &&
+                    binder.Syntax == syntax &&
+                    binder.CollectionType.OriginalDefinition.Equals(targetType.OriginalDefinition, TypeCompareKind.AllIgnoreOptions))
+                {
+                    return true;
+                }
+                current = current.Next;
+            } while (current is { });
+
+            return false;
+        }
+
+        internal PropertySymbol? GetCollectionExpressionApplicableIndexerIfEnabled(SyntaxNode syntax, TypeSymbol targetType, TypeSymbol elementType, BindingDiagnosticBag diagnostics)
+        {
+            if (object.ReferenceEquals(elementType.OriginalDefinition, Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_KeyValuePair_KV)) &&
+                Compilation.IsFeatureEnabled(MessageID.IDS_FeatureDictionaryExpressions))
+            {
+                return GetCollectionExpressionApplicableIndexer(syntax, targetType, elementType, diagnostics);
+            }
+            return null;
+        }
+
+        internal PropertySymbol? GetCollectionExpressionApplicableIndexer(SyntaxNode syntax, TypeSymbol targetType, TypeSymbol elementType, BindingDiagnosticBag diagnostics)
+        {
+            bool isKeyValuePair = ConversionsBase.IsKeyValuePairType(Compilation, elementType, out var keyType, out var valueType);
+            Debug.Assert(keyType is { });
+            Debug.Assert(valueType is { });
+
+            var receiver = new BoundValuePlaceholder(syntax, targetType);
```

Comment:
Not returning this up as an explicit placeholder makes me uneasy. Where do we record this for later replacement?

Reply:
We're not returning the `BoundIndexerAccess` so we won't need to replace the placeholder.